docs/proposals/backend-support-appset.md

```diff
@@ -0,0 +1,226 @@
+
+8. Once all the checks have passed, apply the ApplicationSet and the Applications, to the namespace.
+
+#### **Should we add a new 'applicationset' RBAC resource?**
+
+* Do we need to add a new RBAC resource for applicationsets, alongside the existing ones? (clusters, projects, applications, repositories, certificates, accounts, gpgkeys)
+Not mandatory for the design explained in this proposal.
+
+* If we DID add a new RBAC resource, this would require application administrators to add a new ApplicationSet resource to the RBAC policy list, alongside their existing Application policies and to keep the ApplicationSet and Application RBAC policy lists in sync. Or, said another way: there would be very few (if any) cases where a user would need to be able to create an Application, but not be able to create an ApplicationSet (and vice versa). When it comes to security, the less moving parts that need to be synchronized, the better (with some minor loss in flexibility, eg the ability to specifically prevent access to ApplicationSet resource).
+
+* If there is no RBAC resource for applicationsets, how do we control access to them?
+Instead of an applicationset rbac resource, we instead examine the Applications that are owned by the applicationset. An ApplicationSet inherits permissions from its generated children.
+Example: if you try to delete an ApplicationSet, we check if you can delete it’s child Applications.
+
+#### **ApplicationSet Deletion Algorithm:**
+
+A user is only able to delete an ApplicationSet if they have permissions to delete all of the Applications managed by the ApplicationSet. This check is performed in ApplicationSet controller, on receiving a delete request via GRPC from API server.
+
+* For each application owned by the ApplicationSet that the user is attempting to delete:
+    * Check if the user has delete permission on the Application
+    * Check if the user has delete permission within the project (?)
+* If the user does NOT have permission on least one of these, the operation should fail.
+* On pass, ApplicationController server deletes (ie kubectl delete) the ApplicationSet resource.
+
+#### **ApplicationSet Update Algorithm:**
+
+A user can only update an ApplicationSet if the user has permission to update all of the Applications currently owned by the ApplicationSet.
+
+* When the user makes a change to an ApplicationSet, we assume that it's possible that the change might affect any or all of the Applications, and thus we require the user to have write access to all of those Applications.
+* We likewise check that the resulting generated Applications are also compliant with the user's permissions.
```

Comment:
We can't know at request-time precisely what applications will eventually be created by this appset. We can only know what applications might _theoretically_ be created.

Reply:
Agreed: the ApplicationSet controller should definitely do these checks each time the ApplicationSet controller Reconciles an ApplicationSet, and if there are any changes, the resulting Applications should be checked. We should add this to the proposal.

Reply:
And actually, we can't even know what applications might theoretically be created, because doing static analysis on an ApplicationSet to determine its potential is challenging.

We may need to restrict users to only creating/modifying ApplicationSets that the ApplicationSet controller can reliably predict their all possible behaviour.

For example:
A) Requiring that security-sensitive Application fields, such `.spec.project`, and `.spec.source.repoURL`, and `.spec.destination`, contain constants (e.g. not be based on parameters generator )
B) Limiting generators to those that do not take external, user-controlled data. (for example, no more Git generator)
- Some combination of A) and B), for example not allowing `.spec.project` to be customized using a value from the Git generator.

Reply:
> we can't always know what Applications might theoretically be created, because doing static analysis on an arbitrary ApplicationSet is challenging.

I think this might actually be possible. I think a simple check for "contains `{{`" in the project field should be enough to tell us the outer boundaries of the appset (either * or a particular project).

> C) Some combination of A) and B), for example not allowing .spec.project to be customized using a value from the Git generator.

Yep, this logic might be difficult to get right, but I think with enough unit tests, anything is possible.